### ISSUE TITLE
[RyuJIT/x86] redundant spill/fill of a long returned from a call

### ISSUE THREAD

**Issue by mskvortsov** [dotnet/coreclr#8748]
The x86 code being generated for `CnsLng1.exe` has inefficiency:
```
; Assembly listing for method BringUpTest:Main():int
; Emitting BLENDED_CODE for generic X86 CPU
; optimized code
; ebp based frame
; partially interruptible
; Final local variable assignments
;
;* V00 loc0         [V00    ] (  0,   0  )    long  ->  zero-ref   
;  V01 rat0         [V01,T01] (  2,   2  )     int  ->  eax         V00.lo(offs=0x00)
;  V02 rat1         [V02,T02] (  2,   2  )     int  ->  edx         V00.hi(offs=0x04)
;  V03 rat2         [V03,T00] (  3,   6  )    long  ->  [ebp-0x08]   multireg-ret
;
; Lcl frame size = 8

G_M32408_IG01:
       55           push     ebp
       8BEC         mov      ebp, esp
       83EC08       sub      esp, 8

G_M32408_IG02:
       FF152C4C2702 call     [BringUpTest:CnsLng1():long]
       8945F8       mov      dword ptr [ebp-08H], eax
       8955FC       mov      dword ptr [ebp-04H], edx
       8B45F8       mov      eax, dword ptr [ebp-08H]
       8B55FC       mov      edx, dword ptr [ebp-04H]
       85D2         test     edx, edx
       750E         jne      SHORT G_M32408_IG04
       83F801       cmp      eax, 1
       7509         jne      SHORT G_M32408_IG04
       B864000000   mov      eax, 100

G_M32408_IG03:
       8BE5         mov      esp, ebp
       5D           pop      ebp
       C3           ret      

G_M32408_IG04:
       B8FFFFFFFF   mov      eax, -1

G_M32408_IG05:
       8BE5         mov      esp, ebp
       5D           pop      ebp
       C3           ret      

; Total bytes of code 51, prolog size 6 for method BringUpTest:Main():int
; ============================================================
```

**Comment by mskvortsov** [comment]
CC @CarolEidt @sivarv 

JIT issues aside, I wonder why is there a difference at the bytecode level between `CnsLng1.exe` I took from the CI release [artifacts](https://ci.dot.net/job/dotnet_coreclr/job/master/job/release_windows_nt_bld/lastSuccessfulBuild/artifact/bin/tests/) and the one I built locally with Roslyn?

CI version:
<pre>
  .method public hidebysig static int32  Main() cil managed
  {
    .entrypoint
    // Code size       16 (0x10)
    .maxstack  2
    .locals init (int64 V_0)
    IL_0000:  call       int64 BringUpTest::CnsLng1()<b>
    IL_0005:  stloc.0
    IL_0006:  ldloc.0</b>
    IL_0007:  ldc.i4.1
    IL_0008:  conv.i8
    IL_0009:  bne.un.s   IL_000e

    IL_000b:  ldc.i4.s   100
    IL_000d:  ret

    IL_000e:  ldc.i4.m1
    IL_000f:  ret
  } // end of method BringUpTest::Main
</pre>
Roslyn version:
<pre>
  .method public hidebysig static int32  Main() cil managed
  {
    .entrypoint
    // Code size       14 (0xe)
    .maxstack  8
    IL_0000:  call       int64 BringUpTest::CnsLng1()
    IL_0005:  ldc.i4.1
    IL_0006:  conv.i8
    IL_0007:  bne.un.s   IL_000c

    IL_0009:  ldc.i4.s   100
    IL_000b:  ret

    IL_000c:  ldc.i4.m1
    IL_000d:  ret
  } // end of method BringUpTest::Main
</pre>

**Comment by mikedn** [comment]
It seems that the CI uses the pre-Roslyn C# compiler.

**Comment by mskvortsov** [comment]
@mikedn I see. My thought was CI builds are deliberately tuned to generate non-optimized code to benchmark the JIT perf in a sense.

And I see the JIT still generates that redundant transfer to memory even with optimized Roslyn code at the input. Seems that `lvaPromoteLongVars` lowers long nodes into loads/stores to a newly created `lclVar`s and they survive till the codegen not being optimized out before.

**Comment by mikedn** [comment]
> My thought was CI builds are deliberately tuned to generate non-optimized code to benchmark the JIT perf in a sense.

Some tests are compiled multiple times with different optimization options (see test projects that have d/do/r/ro suffixes) but this is not the case with codegen bring up tests, they have optimization enabled. You can check the presence of the following assembly attribute in the test exe:
```C#
[assembly: Debuggable(DebuggableAttribute.DebuggingModes.IgnoreSymbolStoreSequencePoints)]
```
Unoptimized exes usually have something like
```C#
[assembly: Debuggable(DebuggableAttribute.DebuggingModes.DisableOptimizations | 
                      DebuggableAttribute.DebuggingModes.IgnoreSymbolStoreSequencePoints | 
                      DebuggableAttribute.DebuggingModes.Default)]
```

> And I see the JIT still generates that redundant transfer to memory even with optimized Roslyn code at the input. Seems that lvaPromoteLongVars lowers long nodes into loads/stores to a newly created lclVars and they survive till the codegen not being optimized out before.

`DecomposeLongs::DecomposeCall` forces the result of the call to be stored in a local variable. AFAIR this was some kind of workaround for an issue but I don't remember what the issue was. Carol/Sivarv should know.

**Comment by mskvortsov** [comment]
Yes, I usually check for the `( 01 00 02 00 00 00 00 00 )` flag presence in `ildasm` output to make sure JIT will work in the opt mode. I meant a bytecode itself could be deliberately made poorly optimized to e.g. check how JIT (running in opt mode) performs on such a contexts.

After I looked through a number the past reviews, I'm almost sure the issue is a well known one. As a matter of fact, Sivarv has already pointed me out to #5890 yesterday.

**Comment by mikedn** [comment]
Hmm, I suppose the issue is something like this:
```C#
l1 = LongCall() + l2;
```
The long addition operation will get decomposed into 2 separate `GT_ADD` nodes and both these nodes need to "use" the call node. Not only that it isn't possible for a node to be used by more than one node but it's not clear how you could specify which half of the long each of the `GT_ADD` nodes use.

**Comment by mskvortsov** [comment]
> it isn't possible for a node to be used by more than one node
@mikedn I don't quite get this. What is the sense of "use" here? Could you elaborate, please?

**Comment by mikedn** [comment]
It is not possible to have IR like (in linear form, I'm not going to draw trees...):
```
t1 = GT_CALL ...
t2 = GT_ADD_LO t1, ...
t3 = GT_ADD_HI t1, ...
```
Every time a node needs to be used in more than one place a local var needs to be introduced so the IR looks like:
```
t1 = GT_CALL ...
lcl1 = GT_STORE_LCLVAR t1
t2 = GT_LCL_VAR lcl1
t2 = GT_ADD_LO t2, ...
t3 = GT_LCL_VAR lcl1
t4 = GT_ADD_HI t3, ...
```

**Comment by mskvortsov** [comment]
Thanks, it's more clear now.

However I probably should not try to get deeper. I stumbled upon the issue while working on #8496 and checking how `long` nodes are implemented for x86.

**Comment by gkhanna79** [comment]
CC @RussKeldorph

**Comment by RussKeldorph** [comment]
@dotnet/jit-contrib
